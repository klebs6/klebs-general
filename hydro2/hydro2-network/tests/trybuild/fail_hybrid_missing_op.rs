// ---------------- [ File: tests/trybuild/fail_hybrid_missing_op.rs ]
// tests/trybuild/fail_hybrid_missing_op.rs

use std::marker::PhantomData;
use hydro2_network_wire_derive::NetworkWire;
use hydro2_operator::*;
use hydro2_operator_derive::*;
use hydro2_network::*;
use hydro2_3p::*;

// We define an operator FooOperator, but do NOT reference it in #[available_operators] => 
// So the expansions that produce references to FooOperatorIO will fail if we build a node with it.

#[derive(Debug, NamedItem, Operator)]
#[operator(
    execute="exec_foo",
    opcode="BasicOpCode::Foo",
    input0="i32",
    output0="i32"
)]
pub struct FooOperator 
{
    name: String,
}

impl FooOperator {
    pub async fn exec_foo(&self, input: &i32) -> NetResult<i32> {
        Ok(input + 1000)
    }
}

#[derive(NetworkWire)]
#[available_operators(
    // We do NOT mention "FooOperator" here => 
    // if we try to build a network with node!(0 => FooOperator), that might fail type-wire checks
    op="BarOperator" // Suppose we define it somewhere else
)]
pub struct MyNetworkWireC {
    _p: PhantomData<i32>,
}

fn main() {
    // Attempt to build a network with node0=FooOperator, but we never added `op="FooOperator"` 
    // => the expansions for the wire type won't have FooOperatorIO => compile error.
    let _node0 = node!(0 => FooOperator { name: "foo".to_string() });
    // ...
}
