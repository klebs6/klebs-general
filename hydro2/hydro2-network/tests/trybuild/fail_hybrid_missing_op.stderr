error[E0412]: cannot find type `BarOperatorIO` in this scope
  --> tests/trybuild/fail_hybrid_missing_op.rs:32:10
   |
32 | #[derive(NetworkWire)]
   |          ^^^^^^^^^^^ not found in this scope
   |
   = note: this error originates in the derive macro `NetworkWire` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `BarOperatorIO` in this scope
  --> tests/trybuild/fail_hybrid_missing_op.rs:32:10
   |
32 | #[derive(NetworkWire)]
   |          ^^^^^^^^^^^ not found in this scope
   |
   = note: this error originates in the derive macro `NetworkWire` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you might be missing a type parameter
   |
38 | pub struct MyNetworkWireC<BarOperatorIO> {
   |                          +++++++++++++++

error[E0412]: cannot find type `BarOperatorOperatorSignature` in this scope
  --> tests/trybuild/fail_hybrid_missing_op.rs:32:10
   |
32 | #[derive(NetworkWire)]
   |          ^^^^^^^^^^^ not found in this scope
   |
   = note: this error originates in the derive macro `NetworkWire` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `BarOperator` in this scope
  --> tests/trybuild/fail_hybrid_missing_op.rs:32:10
   |
32 | #[derive(NetworkWire)]
   |          ^^^^^^^^^^^ not found in this scope
   |
   = note: this error originates in the derive macro `NetworkWire` (in Nightly builds, run with -Z macro-backtrace for more info)
